source/common/http/async_client_impl.cc

```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
       dispatcher_(dispatcher) {}
 
 AsyncClientImpl::~AsyncClientImpl() {
-  while (!active_requests_.empty()) {
-    active_requests_.front()->failDueToClientDestroy();
+  while (!active_streams_.empty()) {
+    active_streams_.front()->reset();
   }
 }
 
 AsyncClient::Request* AsyncClientImpl::send(MessagePtr&& request, AsyncClient::Callbacks& callbacks,
                                             const Optional<std::chrono::milliseconds>& timeout) {
-  std::unique_ptr<AsyncRequestImpl> new_request{
-      new AsyncRequestImpl(std::move(request), *this, callbacks, timeout)};
+  AsyncRequestImpl* async_request =
+      new AsyncRequestImpl(std::move(request), *this, callbacks, timeout);
+  std::unique_ptr<AsyncStreamImpl> new_request{async_request};
 
   // The request may get immediately failed. If so, we will return nullptr.
-  if (!new_request->complete_) {
-    new_request->moveIntoList(std::move(new_request), active_requests_);
-    return active_requests_.front().get();
+  if (!new_request->complete()) {
```

Comment:
If I'm reading the code correctly, I think you actually want to check for remote_close_ here now, not complete_, or, arguably better, you need to set local_close_ in resetStream(). 

The reason for this is that the router might respond right away. If it responds before the request is complete, it will reset. In this case, you should return a null handle.